File: core/src/main/java/net/techcable/sonarpet/nms/entity/generators/CustomMotionPetGenerator.java
package net.techcable.sonarpet.nms.entity.generators;

import net.techcable.sonarpet.utils.Versioning;
import net.techcable.sonarpet.utils.reflection.MinecraftReflection;

import org.objectweb.asm.Type;

import static org.objectweb.asm.Opcodes.*;

/**
 * Some entity specific logic is kept in the living update/motion method, like enderdragon destruction.
 * This generator overrides the default (entity-specific) update logic, with a minimalistic alternative.
 */
public class CustomMotionPetGenerator extends EntityPetGenerator {
    public CustomMotionPetGenerator(Type currentType, Class<?> hookClass, Class<?> entityClass) {
        super(currentType, hookClass, entityClass);
    }

    @Override
    protected void generate0() {
        super.generate0();
        final String livingUpdateMethod = Versioning.NMS_VERSION.getObfuscatedMethod("LIVING_UPDATE_METHOD");
        // Nope, don't move funky
        generator.generateMethod(
                (generator) -> {
                    // TODO: Fix the bug with these pets not actually moving (maybe something to do with a missing setPosition?)
                    // this.doTick()
                    generator.loadThis();
                    generator.invokeVirtual(
                            "doTick",
                            ENTITY_LIVING_TYPE,
                            Type.VOID_TYPE
                    );
                    final String sidewaysMotionField = Versioning.NMS_VERSION.getObfuscatedField("ENTITY_SIDEWAYS_MOTION_FIELD");
                    final String forwardMotionField = Versioning.NMS_VERSION.getObfuscatedField("ENTITY_FORWARD_MOTION_FIELD");
                    // this.moveStrafing *= 0.98
                    // this.moveForward *= 0.98
                    for (String fieldName : new String[] {sidewaysMotionField, forwardMotionField}) {
                        generator.loadThis();
                        generator.visitInsn(DUP);
                        generator.getField(ENTITY_LIVING_TYPE, fieldName, Type.FLOAT_TYPE);
                        generator.visitLdcInsn(0.98f);
                        generator.visitInsn(FMUL);
                        generator.putField(ENTITY_LIVING_TYPE, fieldName, Type.FLOAT_TYPE);
                    }
                    // this.move(this.moveStrafing, this.moveForward)
                    generator.loadThis();
                    generator.visitInsn(DUP);
                    generator.getField(ENTITY_LIVING_TYPE, sidewaysMotionField, Type.FLOAT_TYPE);
                    generator.loadThis();
                    generator.getField(ENTITY_LIVING_TYPE, forwardMotionField, Type.FLOAT_TYPE);
                    generator.invokeVirtual(
                            Versioning.NMS_VERSION.getObfuscatedMethod("ENTITY_MOVE_METHOD"),
                            ENTITY_LIVING_TYPE,
                            Type.VOID_TYPE,
                            Type.FLOAT_TYPE,
                            Type.FLOAT_TYPE
                    );
                },
                ACC_PUBLIC,
                livingUpdateMethod,
                Type.VOID_TYPE
        );
    }

    //
    // Type constants
    //
    private static final Type ENTITY_LIVING_TYPE = Type.getType(MinecraftReflection.getNmsClass("EntityLiving"));
}

File: core/src/main/kotlin/net/techcable/sonarpet/nms/entity/generators/CustomMotionPetGenerator.kt
package net.techcable.sonarpet.nms.entity.generators

import com.dsh105.echopet.compat.api.plugin.IEchoPetPlugin
import com.google.common.collect.ImmutableList
import net.techcable.sonarpet.nms.*
import net.techcable.sonarpet.utils.Versioning.*
import net.techcable.sonarpet.utils.invokeVirtual
import net.techcable.sonarpet.utils.plus
import org.objectweb.asm.Opcodes.*
import org.objectweb.asm.Type
import org.objectweb.asm.Type.*

/**
 * Some entity specific logic is kept in the living update/motion method, like enderdragon destruction.
 * This generator overrides the default (entity-specific) update logic, with a minimalistic alternative.
 */
class CustomMotionPetGenerator(plugin: IEchoPetPlugin, currentType: Type, hookClass: Class<*>, entityClass: Class<*>) : EntityPetGenerator(plugin, currentType, hookClass, entityClass) {
    override val generatedMethods: ImmutableList<GeneratedMethod>
        get() = super.generatedMethods + GeneratedMethod(NMS_VERSION.livingUpdateMethod) {
            // TODO: Fix the bug with these pets not actually moving (maybe something to do with a missing setPosition?)
            // this.doTick()
            loadThis()
            invokeVirtual("doTick", ENTITY_LIVING_TYPE, VOID_TYPE)
            // this.moveStrafing *= 0.98
            // this.moveForward *= 0.98
            // this.moveUpwards *= 0.9
            val movementUpdates = mutableMapOf(
                    NMS_VERSION.sidewaysMotionField to 0.98f,
                    NMS_VERSION.forwardMotionField to 0.98f
            )
            // NOTE: Upwards motion is only present post 1.12 and is multiplied by 0.9 not 0.98
            NMS_VERSION.upwardsMotionField?.let { movementUpdates.put(it, 0.9f) }
            for ((fieldName, coefficient) in movementUpdates) {
                loadThis()
                visitInsn(DUP)
                getField(ENTITY_LIVING_TYPE, fieldName, FLOAT_TYPE)
                visitLdcInsn(coefficient)
                visitInsn(FMUL)
                putField(ENTITY_LIVING_TYPE, fieldName, FLOAT_TYPE)
            }
            // this.move(this.moveStrafing, this.moveForward, this.moveUpwards), moveUpwards missing before 1.12
            loadThis()
            visitInsn(DUP)
            getField(ENTITY_LIVING_TYPE, NMS_VERSION.sidewaysMotionField, FLOAT_TYPE)
            loadThis()
            getField(ENTITY_LIVING_TYPE, NMS_VERSION.forwardMotionField, FLOAT_TYPE)
            NMS_VERSION.upwardsMotionField?.let {
                loadThis()
                getField(ENTITY_LIVING_TYPE, it, FLOAT_TYPE)
            }
            invokeVirtual(
                    name = NMS_VERSION.entityMoveMethodName,
                    ownerType = ENTITY_LIVING_TYPE,
                    parameterTypes = NMS_VERSION.entityMoveMethodParameters.toList()
            )
        }
}